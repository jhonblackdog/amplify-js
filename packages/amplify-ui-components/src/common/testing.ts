export const pixelThreshold = 10;
