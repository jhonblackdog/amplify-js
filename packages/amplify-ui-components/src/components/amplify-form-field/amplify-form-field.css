amplify-form-field input {
  padding: 5px;
  width: 100%;
  max-width: 300px;
  box-sizing: border-box;
  border: 1px gray solid;
}

amplify-form-field input:focus {
  outline: none;
  border-color: orange;
}

amplify-form-field .label {
  display: block;
  font-size: 0.9em;
  font-weight: bold;
  padding-bottom: 0.5em;
}

amplify-form-field .description {
  font-size: 0.8em;
  padding-top: 0.5em;
}
