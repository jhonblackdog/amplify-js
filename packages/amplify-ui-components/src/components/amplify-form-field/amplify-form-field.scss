.form-field {
  margin-bottom: 15px;
}

.form-field-label {
  display: block;
  font-size: 0.9em;
  padding-bottom: 0.5em;
}

.form-field-description {
  font-size: 0.8em;
  padding-top: 0.5em;
}
