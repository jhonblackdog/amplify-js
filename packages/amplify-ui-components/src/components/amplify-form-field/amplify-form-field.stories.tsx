import { storiesOf } from '@storybook/html';
import { text as textKnob } from '@storybook/addon-knobs';

storiesOf('amplify-form-field', module).add('with knobs', () => {
  const label = textKnob('Label', 'Label placeholder');
  const description = textKnob('Description', 'Description placeholder');
  const type = textKnob('Type', 'text');
  return `<amplify-form-field field-id="id" label="${label}" description="${description}" type="${type}"/>`;
});

storiesOf('amplify-form-field', module).add('password example', () => {
  return `<amplify-form-field field-id="id" label="Enter your password" description="Passwords should be kept safe!" type="password"/>`;
});
