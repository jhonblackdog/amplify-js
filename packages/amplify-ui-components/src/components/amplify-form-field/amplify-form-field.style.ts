import { css } from 'emotion';

export const formField = css`
  input {
    padding: 5px;
    width: 100%;
    max-width: 300px;
    box-sizing: border-box;
    border: 1px gray solid;
  }

  input:focus {
    outline: none;
    border-color: orange;
  }
`;

export const formLabel = css`
  .label {
    display: block;
    font-size: 0.9em;
    font-weight: bold;
    padding-bottom: 0.5em;
  }
`;

export const formDescription = css`
  .description {
    font-size: 0.8em;
    padding-top: 0.5em;
  }
`;
