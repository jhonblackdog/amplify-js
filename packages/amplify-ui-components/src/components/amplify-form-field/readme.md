# amplify-form-field



<!-- Auto Generated Below -->


## Properties

| Property        | Attribute     | Description                                                                                                                                        | Type                                                                                            | Default     |
| --------------- | ------------- | -------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------- | ----------- |
| `description`   | `description` | The text of the description.  Goes between the label and the input.                                                                                | `string`                                                                                        | `undefined` |
| `fieldId`       | `field-id`    | The ID of the field.  Should match with its corresponding input's ID.                                                                              | `string`                                                                                        | `undefined` |
| `hint`          | `hint`        | The text of a hint to the user as to how to fill out the input.  Goes just below the input.                                                        | `string`                                                                                        | `undefined` |
| `label`         | `label`       | The text of the label.  Goes above the input. Ex: 'First name'                                                                                     | `string`                                                                                        | `undefined` |
| `onInputChange` | --            | The callback, called when the input is modified by the user.                                                                                       | `(inputEvent: Event) => void`                                                                   | `undefined` |
| `override`      | `override`    | (Optional) Override default styling                                                                                                                | `boolean`                                                                                       | `false`     |
| `placeholder`   | `placeholder` | (Optional) The placeholder for the input element.  Using hints is recommended, but placeholders can also be useful to convey information to users. | `string`                                                                                        | `''`        |
| `type`          | `type`        | The input type.  Can be any HTML input type.                                                                                                       | `"date" \| "email" \| "number" \| "password" \| "search" \| "tel" \| "text" \| "time" \| "url"` | `'text'`    |


## Dependencies

### Used by

 - [amplify-examples](../amplify-examples)
 - [amplify-sign-in-password-field](../amplify-sign-in-password-field)
 - [amplify-sign-in-username-field](../amplify-sign-in-username-field)

### Depends on

- [amplify-label](../amplify-label)
- [amplify-input](../amplify-input)
- [amplify-hint](../amplify-hint)

### Graph
```mermaid
graph TD;
  amplify-form-field --> amplify-label
  amplify-form-field --> amplify-input
  amplify-form-field --> amplify-hint
  amplify-examples --> amplify-form-field
  amplify-sign-in-password-field --> amplify-form-field
  amplify-sign-in-username-field --> amplify-form-field
  style amplify-form-field fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
