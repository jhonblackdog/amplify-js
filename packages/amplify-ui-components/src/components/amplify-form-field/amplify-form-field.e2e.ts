import { newE2EPage } from '@stencil/core/testing';
import { pixelThreshold, selectors } from '../../common/testing';

describe('amplify-form-field', () => {
  it('renders', async () => {
    const page = await newE2EPage();

    await page.setContent(
      `<amplify-form-field field-id="id" label="Label test" description="Description test" hint="Hint test" placeholder="Placeholder test"></amplify-form-field>`,
    );
    const screenshot = await page.compareScreenshot('Amplify Form Field', {fullPage: true});
    expect(screenshot).toMatchScreenshot({ allowableMismatchedPixels: pixelThreshold });

    const fieldElement = await page.find('amplify-form-field');
    expect(fieldElement).not.toBeNull();

    const labelElement = await page.find('label');
    expect(labelElement).toEqualText('Label test');
    expect(labelElement).toEqualAttribute('for', 'id');

    const descriptionElement = await page.find(selectors.formFieldDescription);
    expect(descriptionElement).toEqualText('Description test');
  });

  it('renders no label or description if none are provided', async () => {
    const page = await newE2EPage();

    await page.setContent(`<amplify-form-field></amplify-form-field>`);

    const labelElement = await page.find('label');
    expect(labelElement).toBeNull();

    const descriptionElement = await page.find(selectors.formFieldDescription);
    expect(descriptionElement).toBeNull();
  });

  it('fires an onInput event when the contents of the box are changed', async () => {
    const page = await newE2EPage();

    await page.setContent(`<amplify-form-field></amplify-form-field>`);

    const func = jest.fn();
    await page.exposeFunction('exposedfunc', func);

    await page.$eval('amplify-form-field', (fieldElement: any) => {
      fieldElement.onInput = this.exposedfunc;
      fieldElement.label = 'adding a label so that the component rerenders';
    });
    await page.waitForChanges();

    const input = await page.find('input');
    await input.press('8');
    expect(func).toBeCalledTimes(1);
    expect(func.mock.calls[0][0].isTrusted).toBe(true);
    const value = await input.getProperty('value');
    expect(value).toBe('8');
  });

  it('can have a checkbox input', async () => {
    const page = await newE2EPage();

    await page.setContent(`<amplify-form-field type='checkbox'></amplify-form-field>`);
    await page.waitForChanges();

    const input = await page.find('input');
    expect(input).toEqualAttribute('type', 'checkbox');
  });

  it('can have a number input', async () => {
    const page = await newE2EPage();

    await page.setContent(`<amplify-form-field type='number'></amplify-form-field>`);
    await page.waitForChanges();

    const input = await page.find('input');
    expect(input).toEqualAttribute('type', 'number');
  });
});
