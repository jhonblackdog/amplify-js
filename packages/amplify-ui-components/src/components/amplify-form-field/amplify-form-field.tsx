import { Component, Prop, h } from '@stencil/core';
import { formFieldLabel, formFieldDescription } from './amplify-form-field.style';
import { TextFieldTypes } from '../../common/types';

@Component({
  tag: 'amplify-form-field',
  shadow: false,
})
export class AmplifyFormField {
  /** The ID of the field.  Should match with its corresponding input's ID. */
  @Prop() fieldId: string;
  /** The text of the label.  Goes above the input. Ex: 'First name' */
  @Prop() label: string | null;
  /** The text of the description.  Goes between the label and the input. */
  @Prop() description: string | null;
  /** The text of a hint to the user as to how to fill out the input.  Goes just below the input. */
  @Prop() hint: string | null;
  /** The input type.  Can be any HTML input type. */
  @Prop() type?: TextFieldTypes = 'text';
  /** The callback, called when the input is modified by the user. */
  @Prop() onInput?: (inputEvent: Event) => void;
  /** (optional) The placeholder for the input element.  Using hints is recommended, but placeholders can also be useful to convey information to users. */
  @Prop() placeholder?: string = '';

  render() {
    return (
      <div>
        {this.label && (
          <div class={formFieldLabel}>
            <amplify-label htmlFor={this.fieldId}>
              {this.label}
            </amplify-label>
          </div>
        )}
        {this.description && (
          <div id={`${this.fieldId}-description`} class={formFieldDescription} data-test="form-field-description">
            {this.description}
          </div>
        )}
        <div>
          <slot name='input'>
            <amplify-input
              fieldId={this.fieldId}
              description={this.description}
              type={this.type}
              onInput={this.onInput}
              placeholder={this.placeholder}
            />
          </slot>
        </div>
        {this.hint && (
          <amplify-hint id={`${this.fieldId}-hint`}>{this.hint}</amplify-hint>
        )}
      </div>
    );
  }
}
