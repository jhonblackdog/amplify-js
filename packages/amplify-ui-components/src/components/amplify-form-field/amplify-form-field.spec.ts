import { newSpecPage } from '@stencil/core/testing';
import { AmplifyFormField } from './amplify-form-field'

describe('amplify-form-field', () => {
  it('renders no label or description if none are provided', async () => {
    const page = await newSpecPage({
      components: [AmplifyFormField],
      html: `<amplify-form-field></amplify-form-field>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders a label and description, if they are provided, and no id', async () => {
    const page = await newSpecPage({
      components: [AmplifyFormField],
      html: `<amplify-form-field label='label' description='description'></amplify-form-field>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('renders with an id, if it is provided', async () => {
    const page = await newSpecPage({
      components: [AmplifyFormField],
      html: `<amplify-form-field label='label' description='description' field-id='id'></amplify-form-field>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it('replaces the input component, if a new one is provided', async () => {
    const page = await newSpecPage({
      components: [AmplifyFormField],
      html: `<amplify-form-field label='label' description='description' field-id='id'><input slot="input" /></amplify-form-field>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
