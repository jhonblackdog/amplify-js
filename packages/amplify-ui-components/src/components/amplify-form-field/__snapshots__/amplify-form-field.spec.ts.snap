// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify-form-field spec: Render logic -> renders a label and description, if they are provided, and no id 1`] = `
<amplify-form-field description="description" label="label">
  <!---->
  <div class="amplify-ui--form-field css-9s9ecg">
    <div class="amplify-ui--form-field-label css-1h8rqh2">
      <amplify-label>
        label
      </amplify-label>
    </div>
    <div class="amplify-ui--form-field-description css-10lf99y" data-test="form-field-description" id="undefined-description">
      description
    </div>
    <div>
      <slot-fb name="input">
        <amplify-input description="description" placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;

exports[`amplify-form-field spec: Render logic -> renders no label or description if none are provided 1`] = `
<amplify-form-field>
  <!---->
  <div class="amplify-ui--form-field css-9s9ecg">
    <div>
      <slot-fb name="input">
        <amplify-input placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;

exports[`amplify-form-field spec: Render logic -> renders with an id, if it is provided 1`] = `
<amplify-form-field description="description" field-id="id" label="label">
  <!---->
  <div class="amplify-ui--form-field css-9s9ecg">
    <div class="amplify-ui--form-field-label css-1h8rqh2">
      <amplify-label htmlfor="id">
        label
      </amplify-label>
    </div>
    <div class="amplify-ui--form-field-description css-10lf99y" data-test="form-field-description" id="id-description">
      description
    </div>
    <div>
      <slot-fb name="input">
        <amplify-input description="description" fieldid="id" placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;

exports[`amplify-form-field spec: Render logic -> replaces the input component, if a new one is provided 1`] = `
<amplify-form-field description="description" field-id="id" label="label">
  <!---->
  <div class="amplify-ui--form-field css-9s9ecg">
    <div class="amplify-ui--form-field-label css-1h8rqh2">
      <amplify-label htmlfor="id">
        label
      </amplify-label>
    </div>
    <div class="amplify-ui--form-field-description css-10lf99y" data-test="form-field-description" id="id-description">
      description
    </div>
    <div>
      <input slot="input">
      <slot-fb hidden="" name="input">
        <amplify-input description="description" fieldid="id" placeholder="" type="text"></amplify-input>
      </slot-fb>
    </div>
  </div>
</amplify-form-field>
`;
